lib/cn/kcompact.c

```diff
@@ -175,176 +79,212 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
     if (w->cw_prog_interval && w->cw_progress)
         tprog = jiffies;
 
-    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    err = bin_heap2_create(w->cw_kvset_cnt, kv_item_compare, &bh);
     if (ev(err))
         return err;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (!more || ev(err))
+    sources = malloc(w->cw_kvset_cnt * sizeof(*sources));
+    if (!sources) {
+        err = merr(ENOMEM);
         goto done;
+    }
 
-new_key:
+    for (uint i = 0; i < w->cw_kvset_cnt; i++)
+        sources[i] = kvset_iter_es_get(w->cw_inputv[i]);
 
-    if (atomic_read(w->cw_cancel_request)) {
-        err = merr(ev(ESHUTDOWN));
+    err = bin_heap2_prepare(bh, w->cw_kvset_cnt, sources);
+    if (ev(err))
         goto done;
-    }
 
-    if (tprog) {
-        u64 now = jiffies;
+    w->cw_stats.ms_srcs = w->cw_kvset_cnt;
+
+    more = bin_heap2_peek(bh, (void **)&curr);
+    while (more) {
+        uint idx;
+        struct kv_iterator *iter;
 
-        if (now - tprog > w->cw_prog_interval) {
-            tprog = now;
-            w->cw_progress(w);
+        if (atomic_read(w->cw_cancel_request)) {
+            err = merr(ESHUTDOWN);
+            goto done;
         }
-    }
 
-    emitted_val = false;
-    horizon = true;
-    emitted_seq = 0;
-    emitted_seq_pt = 0;
-
-    dbg_prev_seq = 0;
-    dbg_prev_src = 0;
-    dbg_nvals_this_key = 0;
-    dbg_dup = false;
-
-get_values:
-    vdata = NULL;
-
-    while (horizon &&
-           kvset_iter_next_vref(
-               w->cw_inputv[curr.src], &curr.vctx, &seq, &vtype, &vbidx, &vboff,
-               &vdata, &vlen, &complen))
-    {
-        bool should_emit = false;
-
-        /* Assertion logic:
-         *   if (dbg_nvals_this_key)
-         *       assert(dbg_prev_seq > seq);
-         */
-        if (HSE_UNLIKELY(dbg_nvals_this_key && dbg_prev_seq <= seq)) {
-            assert(0);
-            seqno_errcnt++;
+        iter = kvset_cursor_es_h2r(curr->src);
+        for (idx = 0; idx < w->cw_kvset_cnt; idx++) {
+            if (w->cw_inputv[idx] == iter)
+                break;
+        }
+        assert(idx < w->cw_kvset_cnt);
+
+        if (tprog) {
+            const u64 now = jiffies;
+
+            if (now - tprog > w->cw_prog_interval) {
+                tprog = now;
+                w->cw_progress(w);
+            }
         }
 
-        dbg_nvals_this_key++;
-        dbg_prev_seq = seq;
+        emitted_val = false;
+        horizon = true;
+        emitted_seq = 0;
+        emitted_seq_pt = 0;
+
+        dbg_prev_seq = 0;
+        dbg_prev_idx = 0;
+        dbg_nvals_this_key = 0;
+        dbg_dup = false;
 
-        if (seq <= w->cw_horizon) {
-            horizon = false;
-            if (pt_set && seq < pt_seq)
-                continue; /* skip value */
+        w->cw_stats.ms_keys_in++;
+        w->cw_stats.ms_key_bytes_in += key_obj_len(&curr->kobj);
```

Comment:
The accounting of these stats seem to have changed leading to the kcompact test failure.